### PR TITLE
Jwt Auth for http handler

### PR DESCRIPTION
**Pull request by youngsofun**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

1. introduce AuthMgr to init and hold extenal Authenticators
    1. not in servers/http, because the clickhouse may support ladp/kebrose later.
2. introduce JwtAuthenticator
    1. using crate `jwt_simple`, which we already used in meta. pure rust and better interface.
    2. implement a simple jwk client.
    3. temporarily limitation (but the interface for extentions is ready)
        1. only support RS256, which is most widely used
        2. only support pub key with jwk， which is more userful, file based pub key need to specify the alg or guess it from pem.
3. http handler auth
    1. support both Basic and JWT
    2. auth done in a middleware and extract UserInfo.
    3. temporarily auth is not forced. if no Credential provided, fall back to root(should check root is `no_password` when root is real user in meta).
    4. temporarily https should be but is not forced too. 
4. test for jwt is coverd by http handler test with jwt.

related but not done in this PR:

1. a real root user.
2. auto refresh of jwks keys
3. more config option for jwt/jwk. (on demand), e.g. currently assume username in standard 'subject' field int the claims.
4. jwt with pub key in local file. (optinal)

## Changelog

- New Feature

## Related Issues

 #3482

## Test Plan

Unit Tests

### CODE REVIEW

**Comment by youngsofun** [comment]
\review @flaneur2020

File: common/meta/types/src/user_auth.rs [hunk 1]
```diff
@@ -163,7 +171,7 @@ impl AuthInfo {
                 hash_value: p,
                 hash_method: t,
             } => t.to_string(p),
-            AuthInfo::None => "".to_string(),
+            AuthInfo::None | AuthInfo::JWT => "".to_string(),
```
Comment (flaneur2020) [inline common/meta/types/src/user_auth.rs:174]: typo? it seems jwt has already pattern matched in an earlier match arm.

File: query/src/servers/http/v1/http_query_handlers.rs [hunk 2]
```diff
@@ -218,14 +219,15 @@ async fn query_page_handler(
 #[poem::handler]
 pub(crate) async fn query_handler(
     sessions_extension: Data<&Arc<SessionManager>>,
+    user_info: Data<&UserInfo>,
```
Comment (flaneur2020) [inline query/src/servers/http/v1/http_query_handlers.rs:222]: 6666

File: query/src/users/auth/jwt.rs [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+                    )),
+                    Some(subject) => Ok(subject),
+                },
+                Err(err) => Err(ErrorCode::AuthenticateFailure(err.to_string())),
+            },
+        }
+    }
+}
+
+mod jwk {
```
Comment (flaneur2020) [inline query/src/users/auth/jwt.rs:62]: move jwk a seperate file? it seems in-file mod is not very common in the current code base

**Comment by flaneur2020** [comment]
jwt intergration with a cleaner auth interface, lgtm!

File: common/meta/types/src/user_auth.rs (hunk 1 above)
Comment (youngsofun) [reply common/meta/types/src/user_auth.rs:174]: @flaneur2020  in two diff functions get_type and get_auth_string

File: query/src/users/auth/jwt.rs (hunk 3 above)
Comment (youngsofun) [reply query/src/users/auth/jwt.rs:62]: ok

Comment (youngsofun) [reply query/src/users/auth/jwt.rs:62]: thank you，I did noticed that, too.
I coded this way for 1. not too many code here,  2.  I want to stress that jwk is sub mod of jwt.

which is better, cc @BohuTANG :

 1.  jwk is not sub mod of jwt ,  might also make sense

- jwt/
   - mod.rs  # empty
   - jwt.rs
   - jwk.rs

 2.   current code base also not put much code in mod.rs?

- jwt/
   - mod.rs  # jwt code
   - jwk.rs

Comment (flaneur2020) [reply query/src/users/auth/jwt.rs:62]:
```
jwt/
mod.rs # empty
jwt.rs
jwk.rs
```

personally i prefer this style, makes mod.rs simple, only `use` statements inside 🤔

File: common/meta/types/src/user_auth.rs (hunk 1 above)
Comment (flaneur2020) [reply common/meta/types/src/user_auth.rs:174]: oh my bad, sorry :joy:

File: query/src/users/auth/jwt.rs (hunk 3 above)
Comment (youngsofun) [reply query/src/users/auth/jwt.rs:62]: 👌

**Pull request merged**